chain/chain/src/chain.rs

```diff
@@ -693,14 +690,22 @@ impl Chain {
         Ok(())
     }
 
-    /// Do Basic validation of a block upon receiving it. Check that header is valid
-    /// and block is well-formed (various roots match).
-    pub fn validate_block(&mut self, block: &Block) -> Result<(), Error> {
-        self.process_block_header(&block.header(), |_| {})?;
-        Self::check_block_validity(self.runtime_adapter.as_ref(), &self.genesis_block(), block)
+    /// Do basic validation of a block upon receiving it. Check that block is
+    /// well-formed (various roots match).
+    pub fn validate_block(&mut self, block: &MaybeValidated<Block>) -> Result<(), Error> {
+        block
+            .validate_with(|block| {
+                Chain::validate_block_impl(
+                    self.runtime_adapter.as_ref(),
+                    &self.genesis_block(),
+                    block,
+                )
+                .map(|_| true)
+            })
```

Comment:
Hm, looking at the call site, it seems what we actually want here is 

```rust
    pub fn validate_with<VE, F>(&self, validator: F) -> Result<(), VE>
    where
        F: FnOnce(&T) -> Result<(), VE>,
    {
        match self.try_validate_with(|it| Ok::<_, std::convert::Infallible>(validator(it))) {
            Ok(it) => it,
            Err(never) => match never {},
        }
    }

    pub fn try_validate_with<VE, E, F>(&self, validator: F) -> Result<Result<(), VE>, E>
    where
        F: FnOnce(&T) -> Result<Result<(), VE>, E>,
    {
        if self.validated.get() {
            Ok(Ok(()))
        } else {
            let res = validator(&self.payload)?;
            self.validated.set(res.is_ok());
            Ok(res)
        }
    }
```

That is, validation is a `-> Result<(), VE>` computation that either asserts that the thing is valid, or returns validation error. Sometimes, the validation process itself is failable -- we fail for unrelated reason before we can conclude if the stuff is valid or not.


That is somewhat slightly-more-advanced-than-usual type-fu, so the current approach is also ok.

Reply:
Yes, this is something I’ve contemplated. The other use case does take advantage of the bool flag so there’s no one simple solution. For now adding `map` felt reasonable enough and least intrusive solution. But if we start using MaybeValidated more we might want to revisit the design.

Reply:
To clarify, I think the other use-case abuses the bool, and encodes `Result<Result<(),()>, Error>` as `Result<bool, Error>`. That is, the semantics of `Err` actually differs.

Reply:
That kinda depends how you think about the operation. Is it a ‘validate and error out if not valid’ operation or a ‘is this payload valid’ question. Though I agree that the naming ‘validate_with’ suggests the first interpretation.